wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java

```diff
@@ -246,32 +202,19 @@ public String toString(M message) {
     long token = reader.beginMessage();
     for (int tag; (tag = reader.nextTag()) != -1;) {
       FieldBinding<M, B> fieldBinding = fieldBindings.get(tag);
-      if (fieldBinding != null) {
-        try {
-          Object value = fieldBinding.singleAdapter.decode(reader);
+      try {
+        if (fieldBinding != null) {
+          Object value = fieldBinding.singleAdapter().decode(reader);
           fieldBinding.value(builder, value);
-        } catch (RuntimeEnumAdapter.EnumConstantNotFoundException e) {
-          // An unknown Enum value was encountered, store it as an unknown field
-          builder.addVarint(tag, e.value);
-        }
-        continue;
-      }
-
-      RegisteredExtension registeredExtension = extensions.get(tag);
-      if (registeredExtension != null) {
-        try {
-          Object value = registeredExtension.adapter.decode(reader);
-          builder.ensureTagMap().add(registeredExtension.extension, value);
-        } catch (RuntimeEnumAdapter.EnumConstantNotFoundException e) {
-          // An unknown Enum value was encountered, store it as an unknown field
-          builder.addVarint(tag, e.value);
+        } else {
+          Extension<?, ?> extension = reader.getExtension(messageType, tag);
+          Object value = extension.getAdapter().decode(reader);
+          builder.ensureTagMap().add(extension, value);
         }
-        continue;
+      } catch (RuntimeEnumAdapter.EnumConstantNotFoundException e) {
+        // An unknown Enum value was encountered, store it as an unknown field
+        builder.setExtension(Extension.unknown(messageType, tag, FieldEncoding.VARINT), e.value);
       }
-
-      FieldEncoding fieldEncoding = reader.peekFieldEncoding();
-      Object value = fieldEncoding.rawProtoAdapter().decode(reader);
-      builder.ensureTagMap().add(tag, fieldEncoding, value);
     }
     reader.endMessage(token);
 
```

Comment:
This file used to be 1000+ lines of complex logic. Now it's like ~80 of simple logic. A long battle, a big win.